wgsl/index.bs

```diff
@@ -4557,11 +4555,282 @@ TODO: *Stub*: Expression evaluation
+
+### Uniformity rules for statements ### {#uniformity-statements}
 
-### Divergence and reconvergence TODO ### {#divergence-reconvergence}
+In order to properly deal with reconvergence we must know when there is a single way for control to flow out of a statement.
+We achieve this by assigning to each statement a set of *behaviors* out of {Return, Break, Continue, Discard, Nothing}.
+Intuitively, this set corresponds to the ways that control-flow can exit the statement.
 
-### Uniformity restrictions TODO ### {#uniformity-restrictions}
+TODO: do we need an extra behavior for possibly infinite loops? I should check whether some form of reconvergence happens after something like `while(tid mod 2) {}`.
+
+If the set of behaviors for a statement has a single element, this means that we either did not diverge within the statement, or we reconverged, and we pick the vertex corresponding to control-flow at the start of the statement as the vertex corresponding to control-flow at the exit of the statement, overriding the other rules below.
+
+The rules for analyzing statements take as argument both the statement itself and the vertex corresponding to control-flow at the beginning of it (which we'll note "CF" below) and return the following:
+
+* A set of behaviors for the statement
+* A vertex corresponding to control-flow at the exit of it
+* A set of new vertices and edges to add to the graph
+
+In the table below, `CF1, S -> CF2, B` means "run the analysis on S starting with control-flow CF1, apply the required changes to the graph, and name the resulting control-flow CF2, and the resulting set of behaviors B".
+Similarly, `CF, E -> CF, V, B` means "run the analysis on expression E, starting with control-flow CF, apply the required changes to the graph, and name the resulting control-flow vertex CF, the resulting value vertex V, and the resulting set of behaviors B" (see next section for the analysis of expressions).
+
+We have a similar set of rules for expressions in left-value positions, that we denote by `CF, E => CF, L, B`. Instead of computing the vertex which corresponds to the uniformity of the value, it computes the vertex which corresponds to the uniformity of the variable we are adressing.
+
+<table class='data'>
+  <caption>Uniformity rules for statements</caption>
+  <thead>
+    <tr><th>Statement<th>New vertices<th>Recursive analyses<th>Behaviors<th>Resulting control-flow vertex<th>New edges
+  </thead>
+  <tr><td class="nowrap">*s1*; *s2*
+      <td>
+      <td class = "nowrap">*CF*, *s1* -> *CF1*, *B1*<br>
+          *CF1*, *s2* -> *CF2*, *B2*
+      <td>*B1* \ {Nothing} U *B2*
+      <td>*CF2*
+      <td>
+  <tr><td class="nowrap">if (*e*) then *s1* else *s2*
+      <td>*CFinside*, *CFend*
+      <td class="nowrap">*CF*, *e* -> *CF'*, *V*, *B*<br>
+          *CFinside*, *s1* -> *CF1*, *B1*<br>
```

Comment:
Sorry, I'm confused here again. We are defining a new vertex "CFinside", but what is it?
This command is supposed to mean "run the analysis on S starting with control-flow CF1, apply the required changes to t...", so it assumes that CF1 is known and given. But here where we are using this, we aren't defining it in any way.

Reply:
CFinside is a freshly allocated node.
It is passed to the recursive analysis so that it can make edges from other nodes to it.
And from the last column you have "*CFinside* -> {*CF'*, *V*}", which adds two edges starting from it.

In terms of implementation, a vertex/node is just an integer. So this entire rule would look something like this (in pseudo-C++ syntax):
```
unsigned CFinside = ++allocatedVerticesCounter;
unsigned CFend = ++allocatedVerticesCounter;
auto (CF', V, B) = analyzeExpr(CF, e);
auto (CF1, B1) = analyzeStmt(CFinside, s1);
auto (CF2, B2) = analyzeStmt(CFinside, s2);
behaviorSet bReturn = union(B, B1, B2);
edges[CFinside].push_back(CF');
edges[CFinside].push_back(V);
edges[CFend].push_back(CF1);
edges[CFend].push_back(CF2);
return (CFend, bReturn);
```
It is probably not perfect C++ (in particular I have not checked the syntax for creating/destructuring tuples, and we probably should expand the edges vector when we increment allocatedVerticesCounter, and we should not use an apostrophe in an identifier), but it should hopefully clarify the meaning of this table.

Did this answer your question? And if so, do you have any advice on how I should present it in the spec?

Reply:
No, I think we should have some wording here to help the reader, like "CFinside is a node corresponding to XXX", and same for all the nodes we are introducing.

Reply:
```
edges[CFinside].push_back(CF');
```

I agree this is necessary, but I can't figure out from the table where this edge comes from. I'd suggest stating it explicitly.